react-app/src/components/PortfolioScreen/PortfolioScreenAPI.ts

```diff
@@ -0,0 +1,81 @@
+import { useCallback, useEffect, useMemo, useState } from "react";
+import { useCosmos } from "../../api/cosmosAPI";
+import { useQueryClient } from "../../providers/QueryClientProvider";
+import { ConnectionStatus, useWallet } from "../../providers/WalletProvider";
+import { translateAddress } from "../../utils/address";
+import { useStaking } from "../../api/stakingAPI";
+import {
+  RequestState,
+  RequestStateError,
+  RequestStateInitial,
+  RequestStateLoaded,
+  RequestStateLoading,
+} from "../../models/RequestState";
+import { Portfolio } from "./PortfolioScreenModel";
+
+type PortfolioRequestState = RequestState<Portfolio | null>;
+
+export function usePortfolioQuery(): PortfolioRequestState {
+  const [loading, setLoading] = useState(false);
+  const [data, setData] = useState<Portfolio>();
+  const [error, setError] = useState<Error>();
+
+  const wallet = useWallet();
+  const cosmosAPI = useCosmos();
+  const staking = useStaking();
+  const { desmosQuery } = useQueryClient();
+
+  const fetchPortfolio = useCallback(async () => {
+    setLoading(true);
+    if (wallet.status !== ConnectionStatus.Connected) {
+      setError(new Error("Wallet not connected!"));
+      return;
+    }
+    try {
+      const [balance, balanceStaked, balanceUnstaking, profile] =
+        await Promise.all([
+          cosmosAPI.getBalance(),
+          cosmosAPI.getBalanceStaked(),
+          staking.getAmountUnstaking(wallet.account.address),
+          desmosQuery.getProfile(
+            translateAddress(wallet.account.address, "desmos")
+          ),
+        ]);
+
+      setData({
+        profile,
+        balance,
+        balanceStaked,
+        balanceUnstaking,
+        address: wallet.account.address,
+      });
+      setLoading(false);
+    } catch (err: unknown) {
+      if (err instanceof Error) {
+        setError(err);
+      }
+      console.log("Failed to handle fetch portfolio error =", err);
+    }
+  }, [wallet, cosmosAPI, staking, desmosQuery, setError]);
+
+  useEffect(() => {
+    fetchPortfolio().catch((err) => {
+      setError(err);
+    });
+  }, [fetchPortfolio]);
+
+  const requestState = useMemo<PortfolioRequestState>(() => {
+    if (data !== undefined) {
+      return RequestStateLoaded(data);
+    }
+    if (error !== undefined) {
+      return RequestStateError(error);
+    }
+    if (loading) {
+      return RequestStateLoading;
+    }
+    return RequestStateInitial;
+  }, [loading, error, data]);
+
+  return requestState;
+}
```

Comment:
i think we don't need to force ourselves to use requestState structure if it's not a graphql api.

Can you check if we will be using the profile in other screens as well? If so I am thinking if we can just put the profile in WalletProvider so we can just fetch it once after we connect the wallet

Reply:
> i think we don't need to force ourselves to use requestState structure if it's not a graphql api.

Wouldn't it produce a more consistent api surface? It still have all the state types (init, loading, loaded & error)

> Can you check if we will be using the profile in other screens as well? If so I am thinking if we can just put the profile in WalletProvider so we can just fetch it once after we connect the wallet

This set of APIs will be reused in the account page, which has exactly the same content as the your portfolio page, except that it will be data of another wallet address. 

I was thinking later this `usePortfolioQuery` hook can be modified to accept a wallet address arg, and be reused in the account page (which is basically the portfolio page with wallet address as url param)

https://www.figma.com/file/SMaKDFUALEopfVeDHcBemr/DAO-likecoin?node-id=101%3A2993